packages/world/src/systems/ProgramSystem.sol

```diff
@@ -1,157 +1,120 @@
+import { ObjectTypes } from "../ObjectTypes.sol";
 import { EntityId } from "../EntityId.sol";
 import { Vec3 } from "../Vec3.sol";
-import { safeCallProgram, callProgramOrRevert } from "../utils/callProgram.sol";
+import { ProgramId } from "../ProgramId.sol";
+import { IAttachProgramHook, IDetachProgramHook, IProgramValidator } from "../ProgramInterfaces.sol";
+import { SAFE_PROGRAM_GAS } from "../Constants.sol";
 
 contract ProgramSystem is System {
-  using WorldResourceIdInstance for ResourceId;
+  function attachProgram(EntityId caller, EntityId target, ProgramId program, bytes calldata extraData) public payable {
+    caller.activate();
+    (, Vec3 targetCoord) = caller.requireConnected(target);
+    target = target.baseEntityId();
 
-  function _requireInterface(address programAddress, bytes4 interfaceId) internal view {
-    require(
-      ERC165Checker.supportsInterface(programAddress, interfaceId),
-      "Program does not implement the required interface"
-    );
-  }
+    // ForceField fragments don't have a position on the grid, so we need to handle them differently
+    // if (objectTypeId == ObjectTypes.ForceFieldFragment) {
+    //   // TODO: figure out proximity checks for fragments
+    // }
 
-  function _callForceFieldProgram(EntityId forceFieldEntityId, EntityId fragmentEntityId, bytes memory data) internal {
-    // We know fragment is active because its forcefield exists, so we can use its program
-    ResourceId fragmentProgram = fragmentEntityId.getProgram();
-    if (fragmentProgram.unwrap() != 0) {
-      callProgramOrRevert(fragmentProgram, data);
-    } else {
-      callProgramOrRevert(forceFieldEntityId.getProgram(), data);
-    }
-  }
+    require(!target.getProgram().exists(), "Existing program must be detached");
 
-  function attachProgram(
-    EntityId callerEntityId,
-    EntityId targetEntityId,
-    ResourceId programSystemId,
-    bytes calldata extraData
-  ) public payable {
-    callerEntityId.activate();
-    (, Vec3 targetCoord) = callerEntityId.requireConnected(targetEntityId);
+    (, bool publicAccess) = Systems._get(program.toResourceId());
+    require(!publicAccess, "Program system must be private");
 
-    EntityId baseEntityId = targetEntityId.baseEntityId();
-    ObjectTypeId objectTypeId = ObjectType._get(baseEntityId);
+    (EntityId validator, ProgramId validatorProgram) = _getValidatorProgram(targetCoord);
 
-    require(baseEntityId.getProgram().unwrap() == 0, "Program already attached");
+    bytes memory validateProgram = abi.encodeCall(
+      IProgramValidator.validateProgram,
+      (caller, validator, target, program, extraData)
+    );
 
-    // ForceField fragments don't have a position on the grid, so we need to handle them differently
-    if (objectTypeId == ObjectTypes.ForceFieldFragment) {
-      // TODO: figure out proximity checks for fragments
-    }
+    // The validateProgram view function should revert if the program is not allowed
```

Comment:
Why is the validateProgram a view function?

Reply:
so that we prevent reentrant calls, and onAttach will be called afterwards anyways so you can do state changes there. Unless we want forcefield to modify state when validating a program, but it felt more natural like this